src/internal/storeComputed.ts

```diff
@@ -0,0 +1,137 @@
+import type { BaseLink, Consumer, RawStore } from './store';
+import { RawStoreFlags, updateLinkProducerValue } from './store';
+import {
+  COMPUTED_ERRORED,
+  COMPUTED_UNSET,
+  RawStoreComputedOrDerived,
+} from './storeComputedOrDerived';
+import { epoch, notificationPhase } from './storeWritable';
+import { activeConsumer, setActiveConsumer, type ActiveConsumer } from './untrack';
+
+export class RawStoreComputed<T>
+  extends RawStoreComputedOrDerived<T>
+  implements Consumer, ActiveConsumer
+{
+  producerIndex = 0;
+  producerLinks: BaseLink<any>[] | null = null;
+  epoch = -1;
+
+  constructor(public computeFn: () => T) {
+    super(COMPUTED_UNSET);
+  }
+
+  override increaseEpoch(): void {
+    // do nothing
+  }
+
+  override updateValue(): void {
+    const flags = this.flags;
+    if (flags & RawStoreFlags.START_USE_CALLED && this.epoch === epoch) {
+      return;
+    }
+    super.updateValue();
+    this.epoch = epoch;
+  }
+
+  override get(): T {
+    if (
+      !activeConsumer &&
+      !notificationPhase &&
+      this.epoch === epoch &&
+      (!(this.flags & RawStoreFlags.HAS_VISIBLE_ONUSE) ||
+        this.flags & RawStoreFlags.START_USE_CALLED)
+    ) {
+      return this.readValue();
+    }
+    return super.get();
+  }
+
+  addProducer<U, L extends BaseLink<U>>(producer: RawStore<U, L>): U {
+    let producerLinks = this.producerLinks;
+    if (!producerLinks) {
+      producerLinks = [];
+      this.producerLinks = producerLinks;
+    }
+    const producerIndex = this.producerIndex;
+    let link = producerLinks[producerIndex] as L | undefined;
+    if (link?.producer !== producer) {
+      if (link) {
+        const endIndex = Math.max(producerIndex + 1, producerLinks.length);
+        producerLinks[endIndex] = link; // push the existing link at the end (to be removed later)
+      }
+      link = producer.registerConsumer(producer.newLink(this));
+    }
+    producerLinks[producerIndex] = link;
+    this.producerIndex = producerIndex + 1;
+    updateLinkProducerValue(link);
+    if (producer.flags & RawStoreFlags.HAS_VISIBLE_ONUSE) {
+      this.flags |= RawStoreFlags.HAS_VISIBLE_ONUSE;
+    }
+    return producer.updateLink(link);
+  }
+
+  override startUse(): void {
+    const producerLinks = this.producerLinks;
+    if (producerLinks) {
+      for (let i = 0, l = producerLinks.length; i < l; i++) {
```

Comment:
Is the `producerLinks` array potentially mutated during `link.producer.unregisterConsumer`, `link.producer.registerConsumer` and so on ? if not, no need to call `producerLinks.length` at every iteration